### PR TITLE
Modify QPACK steam and instruction terminology

### PR DESCRIPTION
**Pull request by LPardue**
This PR is more controversial than I first anticipated. It is intended to address #2339.

QPACK uses the term instruction extensively section 4 goes into great detail about all the different types. However, the instructions have been presented in terms of what streams they appear on. Which makes things a little unclear, and verbose, sometimes. I think it also conflates the idea of the true QPACK unframed encoder and decoder streams, with header block messages (sets of instructions).

So this change explicitly creates the terms encoder instruction, decoder instruction and header block instruction. It also removes a lot of the reference to request and push streams.

I'm not 100% happy with the net outcome of these changes but posting the PR for feedback.

### CODE REVIEW

File: draft-ietf-quic-qpack.md [hunk 1]
```diff
@@ -713,16 +722,15 @@ references (see {{reference-tracking}}).  Changing the capacity of the dynamic
...
-## Decoder Stream
+## Decoder Instructions {#decoder-instructions}
 
-The decoder stream carries information used to ensure consistency of the dynamic
-table. Information is sent from the decoder to the encoder; that is, the server
-informs the client about the processing of the client's header blocks and table
-updates, and the client informs the server about the processing of the server's
-header blocks and table updates.
+Decoder instructions provide information used to ensure consistency of the
+dynamic table. They are sent from the decoder to the encoder on a decoder stream;
```
Comment (MikeBishop) [inline draft-ietf-quic-qpack.md]: Need to rewrap this paragraph -- your build break is an overlong line here.

File: draft-ietf-quic-qpack.md [hunk 2]
```diff
@@ -580,40 +581,48 @@ prefix integer.  The remainder of the string literal is unmodified.
...
-## Stream Types
+## Instructions
 
-QPACK instructions occur in three locations, each of which uses a separate
-instruction space:
+There are three separate QPACK instruction spaces. Encoder instructions
+({{encoder-instructions}}) carry table updates, decoder instructions
+({{decoder-instructions}}) carry acknowledgments of table modifications and
+header processing, and header block instructions ({{header-block-instructions}})
+reference the QPACK table state.
```
Comment (martinthomson) [inline draft-ietf-quic-qpack.md]: Isn't the main purpose of header block instructions to cause header fields to be emitted?

File: draft-ietf-quic-qpack.md [hunk 3]
```diff
@@ -580,40 +581,48 @@ prefix integer.  The remainder of the string literal is unmodified.
...
-This section describes the instructions which are possible on each stream type.
+ - A decoder stream is a unidirectional stream of type `0x68` (ASCII 'h').
+   It carries an unframed sequence of decoder instructions from decoder
+   to encoder.
 
-## Encoder Stream {#encoder-stream}
+<!-- s/exactly/no more than/  ? -->
+HTTP/3 endpoints contain a QPACK encoder and decoder. Each endpoint MUST
+initiate a single encoder stream and decoder stream. Receipt of a second
+instance of either stream type MUST treat it as a connection error of type
```
Comment (martinthomson) [inline draft-ietf-quic-qpack.md]: Check grammar: "An endpoint that receives a second" or "MUST be treated as"

File: draft-ietf-quic-qpack.md [hunk 4]
```diff
@@ -811,11 +819,13 @@ An encoder cannot infer from this instruction that any updates to the dynamic
 table have been received.
 
 
-## Request and Push Streams
+## Header Block Instructions
+
+HTTP/3 endpoints convert header lists to headers blocks and exchange them inside
+HEADERS and PUSH_PROMISE frames. Header block instructions reference the static
+table, or dynamic table in a particular state without modifying it.
```
Comment (martinthomson) [inline draft-ietf-quic-qpack.md]: Same comment as before.  The primary purpose is to convey the header field values.  That they reference various tables to do so is secondary.

Comment (LPardue) [reply draft-ietf-quic-qpack.md]: Hmm. I think I agree with your position, which is why I have some struggle with the text as it currently stand. The changes here try to carry the existing QPACK position and text where possible. I'll take editor(s) cue on whether to broaden the changes in this PR.

AFAICT the QPACK approach was to discuss instructions and how they relate to the dynamic table. In that sense, this paragraph provides a clear statement that these instructions do not mutate.

File: draft-ietf-quic-qpack.md (hunk 3 above)
Comment (LPardue) [reply draft-ietf-quic-qpack.md]: Ack

File: draft-ietf-quic-qpack.md (hunk 4 above)
Comment (martinthomson) [reply draft-ietf-quic-qpack.md]: I'm OK with that aspect of this.  However, I would first say what these are for, then explain that they don't refer to the tables, but don't modify them.

File: draft-ietf-quic-qpack.md [hunk 5]
```diff
@@ -580,40 +581,48 @@ prefix integer.  The remainder of the string literal is unmodified.
...
-## Stream Types
+## Instructions
 
-QPACK instructions occur in three locations, each of which uses a separate
-instruction space:
+There are three separate QPACK instruction spaces. Encoder instructions
+({{encoder-instructions}}) carry table updates, decoder instructions
+({{decoder-instructions}}) carry acknowledgments of table modifications and
+header processing, and header block instructions ({{header-block-instructions}})
+compress a header list by referring to the QPACK table state.
```
Comment (MikeBishop) [inline draft-ietf-quic-qpack.md]: Well, they don't really *compress* the header list, they are the compression output.  Perhaps "reconstruct" a header list?

File: draft-ietf-quic-qpack.md [hunk 6]
```diff
@@ -1128,7 +1141,7 @@ The entries in the following table are registered by this document.
 
 ## Stream Type Registration
 
-This document creates two new settings in the "HTTP/3 Stream Type" registry
+This document creates two new streams in the "HTTP/3 Stream Type" registry
```
Comment (MikeBishop) [inline draft-ietf-quic-qpack.md]: Perhaps "types" or "entries"?

Comment (LPardue) [reply draft-ietf-quic-qpack.md]: The change addressed a bug in the spec. I'm happy to use entries but would do so consistently through the document (settings, errors, stream)

File: draft-ietf-quic-qpack.md (hunk 5 above)
Comment (LPardue) [reply draft-ietf-quic-qpack.md]: Well, an encoder view of the header block is not to reconstruct the list. 

The initial text simply stated that these instructions refer to tables. Adding flavour seems to complicate the matter, unless we think spelling out verbatim is useful. 

Would something like "header block instructions convey an encoded representation of a header list by referring to the QPACK table state" suffice?

Comment (MikeBishop) [reply draft-ietf-quic-qpack.md]: WFM.

File: draft-ietf-quic-qpack.md (hunk 6 above)
Comment (MikeBishop) [reply draft-ietf-quic-qpack.md]: Well, arguably, the other two actually are creating new settings and error types.  But yes, if you want to fix the bug and remain consistent, this should probably be "types".

File: draft-ietf-quic-qpack.md [hunk 7]
```diff
@@ -580,40 +581,49 @@ prefix integer.  The remainder of the string literal is unmodified.
...
+   It carries an unframed sequence of encoder instructions from encoder
+   to decoder.
 
-This section describes the instructions which are possible on each stream type.
+ - A decoder stream is a unidirectional stream of type `0x68` (ASCII 'h').
+   It carries an unframed sequence of decoder instructions from decoder
+   to encoder.
+
+<!-- s/exactly/no more than/  ? -->
+HTTP/3 endpoints contain a QPACK encoder and decoder. Each endpoint MUST
```
Comment (afrind) [inline draft-ietf-quic-qpack.md]: I think I'll change this to MAY in a subsequent PR

File: draft-ietf-quic-qpack.md (hunk 4 above)
Comment (afrind) [reply draft-ietf-quic-qpack.md]: I'm OK with the text here.

**Pull request merged**